### ISSUE TITLE
Step to create a custom NER pipeline 

### ISSUE THREAD

**Issue by 78guigui95** [explosion/spaCy#3518]
Hello,

I would like to create a custom french NER pipeline from scratch. However I'm not sure about how to do. 
I want to use my own word vectors ( trained with gensim) and add 2 modules to my pipeline : `EntityRuler` and `EntityRecognizer.`

This is what i have understood from the documentation:

- First i have to use `init-model` to create a blank model with my word vectors and then load it:
`python -m spacy init-model en ./data/spacy.word2vec.model --vectors-loc word2vec.txt.gz`
`nlp = spacy.load( './data/spacy.word2vec.model/')`

- Then i have to add the `EntityRuler`
- And the last step is to Train my `EntityRecognizer.`

I have many questions:

- Do i miss something ? is there any other step i forgot ? 
- Does it help EntityRecognizer to import my own word vectors ? And does it help to generalize the NER ? 
 

Thank you very much for your hep.

Best regards.

**Comment by ines** [comment]
> Does it help EntityRecognizer to import my own word vectors ? And does it help to generalize the NER ?

Yes, especially for very specific domains, providing your own vectors can give you a nice boost in accuracy. Just make sure that you're always using the same vectors during training and at runtime. You might also want to try out the new `spacy pretrain`.

> Do i miss something ? is there any other step i forgot ?

Yes, this makes sense. You don't need the `EntityRuler` in the pipeline when you train the model – it's a purely rule-based component so you can also add it to your trained model afterwards.

If you have a lot of rules already, one strategy could be to start off with *only* the `EntityRuler` and see how far it gets you. This will also be very helpful to find out what entities your rules cover easily, and what entities are tricky because they're ambiguous and very context-dependent.

Do you already have labelled data for the entity categories you want to train? If not, you definitely want to be spending some time designing the label scheme. What categories do you need? Do they make sense? Are they easy to learn and easy to label? You also want to have a good evaluation set in place that you can evaluate the rules and the statistical model against. It might not be the most exciting part, but having a good evaluation in place will be incredibly helpful because you'll always know whether (and how) your model is improving and you'll have numbers to back that up.

**Comment by 78guigui95** [comment]
Thanks a lot for you clear and helpful reply ! 

Yes i have labelled data already. I want to train the model to recognize generics locations like 'street', 'cinema', 'shop'.... 
I'm not sure to understand how `EntityRuler` works : if i use it alone, will my model be able to generalize to similar context and locations, or do i need `EntityRecognizer` for that ? 
From the documentation i have understood that `EntityRuler` will only match the exact same words i give him. 

Thanks again !

**Comment by ines** [comment]
> From the documentation i have understood that EntityRuler will only match the exact same words i give him.

Yes, exactly. It also lets you write more abstract token patterns (just like the `Matcher`) – for example, using token attributes like `LOWER` (for case-insensitive matches) or other linguistic features (e.g. only match that string if it's a noun).

So if you only use rules, the model won't be able to generalise, no – but you can use the rules to cover the cases that are 100% unambiguous. For example, `"Apple, Inc."` will likely always be the company. So if you add the entity ruler *before* the statistical entit recognizer, the model will accept the pre-defined entity (and won't predict only "Apple" in that case).

I mostly suggested the rules-only approach as a quick experiment to get a better feeling for your data: you can give the entity ruler a few hundred or thousand exact matches, run it over your evaluation data and calculate the accuracy. Let's say the accuracy is 75% – this gives you a baseline that your statistical model *has to* beat.

**Comment by aschmu** [comment]
Hello,

@ines sorry for reopening this issue, but I'm wondering if you have any experiments comparing
- using custom word vectors (in which case the NER pipeline is trained from scratch)
- vs fine-tuning a pre-trained model (taking into account catastrophic forgetting)
In the latter case, I understand we don't need as many dataset specific examples since the model is benefiting from pre-training. In the former case I suspect custom vectors could help weed out some false positives (e.g. capitalized words in customer complaints) but I'm not sure how much more training data we need in order to reach good performance. I'm contemplating using custom embeddings but I fear the potential advantages would be offset by how much more manual labelling I'd need.
